packages/file_selector/file_selector_windows/lib/src/dart_file_selector_api.dart

```diff
@@ -0,0 +1,328 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+import 'package:file_selector_windows/src/messages.g.dart';
+import 'package:flutter/cupertino.dart';
+import 'package:win32/win32.dart';
+
+import 'dart_file_dialog.dart';
+import 'dart_file_open_dialog_api.dart';
+import 'dart_shell_item_api.dart';
+
+/// Dart native implementation of FileSelectorAPI
+class DartFileSelectorAPI extends FileDialog {
+  /// We need the file to exist. This value defaults to `false`.
+  DartFileSelectorAPI(
+      [FileOpenDialogAPI? fileOpenDialogAPI, ShellItemAPI? shellItemAPI])
+      : super() {
+    _fileOpenDialogAPI = fileOpenDialogAPI ?? FileOpenDialogAPI();
+    _shellItemAPI = shellItemAPI ?? ShellItemAPI();
+  }
+
+  late FileOpenDialogAPI _fileOpenDialogAPI;
+  late ShellItemAPI _shellItemAPI;
+
+  /// Returns a directory path from user selection.
+  String? getDirectoryPath({
+    String? initialDirectory,
+    String? confirmButtonText,
+  }) {
+    fileMustExist = true;
+    final SelectionOptions selectionOptions = SelectionOptions(
+        allowMultiple: false, selectFolders: true, allowedTypes: <TypeGroup>[]);
+    return _getDirectory(
+        initialDirectory: initialDirectory,
+        confirmButtonText: confirmButtonText,
+        selectionOptions: selectionOptions);
+  }
+
+  /// Returns a full path, including file name and it's extension, from user selection.
+  String? getSavePath({
+    String? initialDirectory,
+    String? confirmButtonText,
+    String? suggestedFileName,
+    SelectionOptions? selectionOptions,
+  }) {
+    fileMustExist = false;
+    final SelectionOptions defaultSelectionOptions = SelectionOptions(
+        allowMultiple: false, selectFolders: true, allowedTypes: <TypeGroup>[]);
+    return _getDirectory(
+        initialDirectory: initialDirectory,
+        confirmButtonText: confirmButtonText,
+        suggestedFileName: suggestedFileName,
+        selectionOptions: selectionOptions ?? defaultSelectionOptions);
+  }
+
+  /// Returns a list of file paths.
+  List<String> getFiles(
+      {String? initialDirectory,
+      String? confirmButtonText,
+      required SelectionOptions selectionOptions}) {
+    fileMustExist = false;
+    int hResult = initializeComLibrary();
```

Comment:
COM initialization must be balanced with a corresponding uninitialization call. If there's an error below - for example if showing the dialog fails - and exception will be thrown and COM won't be uninitialized. I'd consider wrapping this in a `try`/`finally` and calling the uninitializing COM in the `finally`.

Reply:
Also, the `fileDialog` must also be released. I'd wrap do that in the `finally` as well.